snuba/utils/streams/batching.py

```diff
@@ -190,75 +227,57 @@ def _run_once(self):
 
         self._handle_message(msg)
 
-    def signal_shutdown(self):
+    def signal_shutdown(self) -> None:
         """Tells the `BatchingKafkaConsumer` to shutdown on the next run loop iteration.
         Typically called from a signal handler."""
         logger.debug("Shutdown signalled")
 
         self.shutdown = True
 
-    def _handle_message(self, msg):
+    def _handle_message(self, msg: Message) -> None:
         start = time.time()
 
         # set the deadline only after the first message for this batch is seen
         if not self.__batch_deadline:
             self.__batch_deadline = self.max_batch_time / 1000.0 + start
 
-        try:
-            result = self.worker.process_message(msg)
-        except Exception:
-            if self.dead_letter_topic:
-                logger.exception("Error handling message, sending to dead letter topic.")
-                self.producer.produce(
-                    self.dead_letter_topic,
-                    key=msg.key(),
-                    value=msg.value(),
-                    headers={
-                        'partition': str(msg.partition()) if msg.partition() else None,
-                        'offset': str(msg.offset()) if msg.offset() else None,
-                        'topic': msg.topic(),
-                    },
-                    on_delivery=self._commit_message_delivery_callback,
-                )
-            else:
-                raise
+        result = self.worker.process_message(msg)
+        if result is not None:
+            self.__batch_results.append(result)
+
+        duration = (time.time() - start) * 1000
+        self.__batch_messages_processed_count += 1
+        self.__batch_processing_time_ms += duration
+        self.__metrics.timing("process_message", duration)
+
+        topic_partition_key = (msg.topic(), msg.partition())
+        if topic_partition_key in self.__batch_offsets:
+            self.__batch_offsets[topic_partition_key].hi = msg.offset()
         else:
-            if result is not None:
-                self.__batch_results.append(result)
-        finally:
-            duration = (time.time() - start) * 1000
-            self.__batch_messages_processed_count += 1
-            self.__batch_processing_time_ms += duration
-            self.__metrics.timing('process_message', duration)
-
-            topic_partition_key = (msg.topic(), msg.partition())
-            if topic_partition_key in self.__batch_offsets:
-                self.__batch_offsets[topic_partition_key][1] = msg.offset()
-            else:
-                self.__batch_offsets[topic_partition_key] = [msg.offset(), msg.offset()]
+            self.__batch_offsets[topic_partition_key] = Offsets(
+                msg.offset(), msg.offset()
```

Comment:
This is not in the finally block anymore. Is this intentional ?
It does not seem to be doing anything worth preserving when this throws, still it changes the state of the object.

Reply:
Semi-intentional. If `process_message` throws, it won't be caught anywhere and will result in that exception being thrown from `run`. I suppose if somebody called `run` again the state would be preserved, but I feel like that's something we shouldn't be doing — it might make sense to go so far as to only allow `run` to be called once in the object's lifetime?

Reply:
I think that could be a good check. You do not need to do it now also because it would be meant to catch a use case that is so broken that it would be pretty hard for someone to build that and expect things will be ok (whoever does that should ask themselves what happens to the message that threw the exception)

Reply:
It's a good observation, I'll do that while I'm in here. The best way is probably to just wrap the `run` method in:

```python
try:
  …  # normal run loop stuff
finally:
  consumer.close()
```

since basically every method will throw a `RuntimeError` after the consumer is closed. Some of the work I have coming up should also make this more explicit.